sdk/go/pulumi/context.go

```diff
@@ -867,14 +838,70 @@ func (ctx *Context) InvokeOutput(
 	output = ctx.newOutput(reflect.TypeOf(output))
 
 	go func() {
+		// Collect dependencies from the DependsOn/DependsOnInput options.
 		invokeOpts := mergeInvokeOptions(options.InvokeOptions...)
-		outProps, deps, err := ctx.invokePackageRaw(tok, args, options.PackageRef, *invokeOpts)
+		deps := []Resource{}
+		depSet := urnSet{} // Only used for `addURNs` below.
+		for _, d := range invokeOpts.DependsOn {
+			if err := d.addURNs(ctx.ctx, depSet, nil); err != nil {
+				return
+			}
+			switch d := d.(type) {
+			case resourceDependencySet:
+				deps = append(deps, d...)
+			case *resourceArrayInputDependencySet:
+				out := d.input.ToResourceArrayOutput()
+				value, known, _, _, err := internal.AwaitOutput(ctx.Context(), out)
+				if err != nil || !known {
+					return
+				}
+				resources, ok := value.([]Resource)
+				if !ok {
+					return
+				}
+				deps = append(deps, resources...)
+			default:
+				// Unreachable.
+				// We control all implementations of dependencySet.
+				contract.Failf("Unknown dependencySet %T", d)
+			}
+		}
+
+		dest := reflect.New(output.ElementType()).Elem()
+
+		// DependsOn for resources is an ordering constraint for register
+		// resource calls. If a resource R1 depends on a resource R2, the
+		// register resource call for R2 will happen after R1. This is ensured
+		// by awaiting the URN for each resource dependency before calling
+		// register resource.
+		//
+		// For invokes, this causes a problem when running under preview. During
+		// preview, register resource immediately returns with the URN, however
+		// this does not tell us if the resource "exists".
+		//
+		// Instead of waiting for the dependency's URN, we wait for the ID. This
+		// tells us that wether a physical resource exists (if the state does
+		// not require a refresh), and we can avoid calling the invoke when it
+		// is unknown.
+		for _, d := range deps {
```

Comment:
I think we need to expand the deps here, so it grabs the children of ComponentResources.

The loop above calling `addURNs` accumulates all the URNs (including those for children), but we don't have the deps themselves.